src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,645 @@
+        precondition!(rel1.is_foreign() != rel2.is_foreign());
+        for [kind1, kind2] in <[AccessKind; 2]>::exhaustive() {
+            for mut state in LocationState::exhaustive() {
+                let protected = true;
+                precondition!(state.perform_access(kind1, rel1, protected).is_ok());
+                precondition!(state.perform_access(kind2, rel2, protected).is_ok());
+                // If these were both allowed, it must have been two reads.
+                assert!(
+                    kind1 == AccessKind::Read && kind2 == AccessKind::Read,
+                    "failed to enforce noalias between two accesses that are not both reads"
+                );
+            }
+        }
+    }
+}
+
+/// We are going to exhaustively test the possibily of inserting
+/// a spurious read in some code.
+///
+/// We choose some pointer `x` through which we want a spurious read to be inserted.
+/// `x` must thus be reborrowed, not have any children, and initially start protected.
+///
+/// To check if inserting a spurious read is possible, we observe the behavior
+/// of some pointer `y` different from `x` (possibly from a different thread, thus
+/// the protectors on `x` and `y` are not necessarily well-nested).
+/// It must be the case that no matter the context, the insertion of a spurious read
+/// through `x` does not introduce UB in code that did not already have UB.
+///
+/// Testing this will need some setup to simulate the evolution of the permissions
+/// of `x` and `y` under arbitrary code. This arbitrary code of course includes
+/// read and write accesses through `x` and `y`, but it must also consider
+/// the less obvious:
+/// - accesses through pointers that are *neither* `x` nor `y`,
+/// - retags of `y` that change its relative position to `x`.
+///
+///
+/// The general code pattern thus looks like
+///     [thread 1]             || [thread 2]
+///                            || y exists
+///     retag x (protect)      ||
+///                      arbitrary code
+///                           read/write x/y/other
+///                        or retag y
+///                        or unprotect y
+///     <spurious read x>      ||
+///                      arbitrary code
+///                           read/write x/y/other
+///                        or retag y
+///                        or unprotect y
+///                        or unprotect x
+///
+/// `x` must still be protected at the moment the spurious read is inserted
+/// because spurious reads are impossible in general on unprotected tags.
+mod spurious_read {
+    use super::*;
+
+    /// Accessed pointer.
+    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    enum PtrSelector {
+        X,
+        Y,
+        Other,
+    }
+
+    /// Relative position of `x` and `y`.
+    /// `y` cannot be a child of `x` because the compiler won't blindly
+    /// insert spurious reads through `x` if it has children.
```

Comment:
```suggestion
    /// `y` cannot be a child of `x` because `x` gets retagged as the first thing in the
    /// pattern, so it cannot have children.
```